IPython/html/static/notebook/js/kernelselector.js

```diff
@@ -154,28 +160,115 @@ define([
         }
     };
 
-    KernelSelector.prototype.set_kernel = function (kernel_name) {
-        /** set the kernel by name, ensuring kernelspecs have been loaded, first */
+    KernelSelector.prototype.set_kernel = function (selected) {
+        /** set the kernel by name, ensuring kernelspecs have been loaded, first 
+        
+        kernel can be just a kernel name, or a notebook kernelspec metadata
+        (name, language, display_name).
+        */
         var that = this;
-        return this.loaded.then(function () {
-            that._set_kernel(kernel_name);
-        });
+        if (typeof selected === 'string') {
+            selected = {
+                name: selected
+            };
+        }
+        if (this._loaded) {
+            this._set_kernel(selected);
+        } else {
+            return this.loaded.then(function () {
+                that._set_kernel(selected);
+            });
+        }
     };
 
-    KernelSelector.prototype._set_kernel = function (kernel_name) {
+    KernelSelector.prototype._set_kernel = function (selected) {
         /** Actually set the kernel (kernelspecs have been loaded) */
-        if (kernel_name === this.current_selection) {
+        if (selected.name === this.current_selection) {
             // only trigger event if value changed
             return;
         }
-        var ks = this.kernelspecs[kernel_name];
+        var kernelspecs = this.kernelspecs;
+        var ks = kernelspecs[selected.name];
+        if (ks === undefined) {
+            var available = _sorted_names(kernelspecs);
+            var matches = [];
+            if (selected.language && selected.language.length > 0) {
+                available.map(function (name) {
+                    if (kernelspecs[name].spec.language.toLowerCase() === selected.language.toLowerCase()) {
+                        matches.push(name);
+                    }
+                });
+            }
+            if (matches.length === 1) {
+                ks = kernelspecs[matches[0]];
+                console.log("No exact match found for " + selected.name +
+                    ", using only kernel that matches language=" + selected.language, ks);
+                this.events.trigger("spec_match_found.Kernel", {
+                    selected: selected,
+                    found: ks,
+                });
+            }
+            // if still undefined, trigger failure event
+            if (ks === undefined) {
+                this.events.trigger("spec_not_found.Kernel", {
```

Comment:
OK, so the `spec_not_found` event really represents two distinct conditions: multiple matches were found, and no matches were found. I'd be inclined to have separate events for those, and present the dialogs slightly differently to make it clearer what's going on. If we do keep a single event, I think it probably warrants an explanatory comment that it can result from two cases.